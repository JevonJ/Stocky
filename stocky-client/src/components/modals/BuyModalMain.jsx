import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Button, Modal, ModalHeader, ModalBody, ModalFooter, ListGroup, ListGroupItem, FormGroup, Form, Input } from 'reactstrap';

class BuyModalMain extends Component {
  constructor() {
    super();
    this.state = {
      quantity: '1',
      totalPrice: '0',
      isWarnVisible: false,
    };
  }

  onChange({ target: { name, value } }) {
    const { stock, stockInfo } = this.props;
    const stockData = stockInfo[stock];
    let isWarnVisible = false;
    if (value > this.validateNofStocks()) {
      value = this.validateNofStocks();
      isWarnVisible = true;
    }
    let totalPrice = value * stockData.currentPrice;
    totalPrice = totalPrice.toFixed(2);

    this.setState({
      quantity: value,
      totalPrice,
      isWarnVisible,
    });
  }

  validateNofStocks() {
    const { user, stock, stockInfo } = this.props;
    const stockData = stockInfo[stock];
    const nos = Math.trunc(user.cash / stockData.currentPrice);
    return (nos);
  }

  buyStock() {
    const {
      user, roomInfo, stock, stockInfo,
    } = this.props;
    const stockData = stockInfo[stock];

    const { room, name, cash } = user;
    const data = {
      room,
      username: name,
      currentCashInHand: cash,
      stockSymbol: stockData.stockSymbol,
      initStockQty: parseInt(this.state.quantity, 10),
      unitPrice: parseInt(stockData.currentPrice, 10),
      round: roomInfo[room].currentRound,
    };

    this.props.socket.emit('purchase_stocks', data);
    this.closeModal();
  }

  closeModal() {
    this.setState({
      quantity: '1',
      totalPrice: '0',
      isWarnVisible: false,
    });
    this.props.toggle();
  }

  render() {
    const { stock, stockInfo } = this.props;
    const { totalPrice, isWarnVisible } = this.state;
    const stockData = stockInfo[stock];

    return (
      <Modal
        id="BuyModal"
        isOpen={this.props.isOpen}
        toggle={() => this.closeModal()}
      >
        <ModalHeader style={{ color: 'black' }} >You are going to buy</ModalHeader>
        <ModalBody style={{ color: 'black' }} className="mbody" >
          <ListGroup>
            <ListGroupItem>Company Symbol: {stockData && stockData.stockSymbol}</ListGroupItem>
            <ListGroupItem>Unit Price: {stockData && stockData.currentPrice}</ListGroupItem>
            <ListGroupItem>
              <FormGroup>
                Number of Stocks
                  <Input
                    type="number"
                    name="qauntity"
                    placeholder="Number of Stocks"
                    max={stockData && this.validateNofStocks()}
                    min={1}
                    value={this.state.quantity}
                    onChange={e => this.onChange(e)}
                  />
              </FormGroup>
              {
                isWarnVisible &&
                <p style={{ color: 'red' }}>Sorry! This is the maximum stocks which you can buy with your remaining cash.</p>
              }
            </ListGroupItem>
            <ListGroupItem>Price: {stockData && (totalPrice === '0' ? stockData.currentPrice : totalPrice)}</ListGroupItem>
          </ListGroup>
        </ModalBody>
        <ModalFooter>
          <Button color="success" onClick={() => this.buyStock()}>Buy</Button>
          <Button color="danger " onClick={() => this.closeModal()}>Not interested</Button>
        </ModalFooter>
      </Modal>
    );
  }
}


const mapStateToProps = ({ user, roomInfo, playerStocks }) => ({
  user,
  roomInfo,
  playerStocks,
});

export default connect(mapStateToProps)(BuyModalMain);
