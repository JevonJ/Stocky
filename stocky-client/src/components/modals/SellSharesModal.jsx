import React, { Component } from 'react';
import { connect } from 'react-redux';
import { Button, Modal, ModalHeader, ModalBody, ModalFooter, ListGroup, ListGroupItem, FormGroup, Form, Input } from 'reactstrap';

class SellSharesModal extends Component {
  constructor() {
    super();
    this.state = {
      quantity: "1",
      totalPrice: "0",
      profitLoss: "0",
      isWarnVisible: false,
    };
  }

  onChange({ target: { name, value } }) {
    const { user, sellStockData } = this.props;
    let isWarnVisible = false;
    if (value >= sellStockData.avlqty) {
      value = sellStockData.avlqty;
      isWarnVisible = true;
    };
    let totalPrice = value * sellStockData.curPrice;
    totalPrice = totalPrice.toFixed(2);

    let profitLoss = totalPrice - (value * sellStockData.oldPrice);
    profitLoss = (Math.abs(profitLoss)).toFixed(2);

    this.setState({
      quantity: value,
      totalPrice,
      profitLoss,
      isWarnVisible,
    });
  }

  sellStock() {
    const { user, roomInfo, sellStockData } = this.props;

    const data = {
      room: user.room,
      username: user.name,
      stockSymbol: sellStockData.symbol,
      stockQty: parseInt(this.state.quantity, 10),
      unitPrice: sellStockData.curPrice,
      round: roomInfo[user.room].currentRound,
    };

    this.props.socket.emit('sell_stocks', data); 
    this.closeModal();
  }

  closeModal() {
    this.setState({
      quantity: "1",
      totalPrice: "0",
      profitLoss: "0",
      isWarnVisible: false,
    });
    this.props.toggle();
  }

  render() {
    const { sellStockData } = this.props;
    const { totalPrice, profitLoss, isWarnVisible } = this.state;

    return (
      <Modal id="SellModal" isOpen={this.props.isOpen} toggle={() => this.closeModal()} className={this.props.className}>
        <ModalHeader style={{ color: 'black' }}>You are going to sell</ModalHeader>
        <ModalBody style={{ color: 'black' }} className="mbody" >
          <ListGroup>
            <ListGroupItem>Company Symbol: {sellStockData && sellStockData.symbol}</ListGroupItem>
            <ListGroupItem>
              <FormGroup>
                Number of Stocks
                <Input
                  type="number"
                  name="qauntity"
                  placeholder="Number of Stocks"
                  max={sellStockData && sellStockData.avlqty}
                  min={1}
                  value={this.state.quantity}
                  onChange={(e) => this.onChange(e)}
                />
              </FormGroup>
              {
                isWarnVisible &&
                <p style={{ color: 'red' }}>Sorry! Cannot exceed your available stock quantity.</p>
              }
            </ListGroupItem>
            <ListGroupItem>Price: {sellStockData && (totalPrice === "0" ? (sellStockData.curPrice).toFixed(2) : totalPrice)}</ListGroupItem>
            <ListGroupItem>{sellStockData && sellStockData.curPrice > sellStockData.oldPrice ? "Profit:" : sellStockData && sellStockData.curPrice < sellStockData.oldPrice ? "Loss:" : "Profit/Loss:"} {sellStockData && (totalPrice === "0" ? (Math.abs(sellStockData.curPrice - sellStockData.oldPrice)).toFixed(2) : profitLoss)}</ListGroupItem>
          </ListGroup>
        </ModalBody>
        <ModalFooter>
          <Button color="success" onClick={() => this.sellStock()}>Sell</Button>
          <Button color="danger " onClick={() => this.closeModal()}>Not interested</Button>
        </ModalFooter>
      </Modal>
    );
  }
}
const mapStateToProps = ({ user, roomInfo, playerStocks }) => ({
  user,
  roomInfo,
  playerStocks,
});
export default connect(mapStateToProps)(SellSharesModal);
