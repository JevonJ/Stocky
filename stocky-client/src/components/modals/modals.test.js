import React from 'react';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

import BuyModalMain from './BuyModalMain';
import SellSharesModal from './SellSharesModal';

const middlewares = [thunk];
const mockStore = configureStore(middlewares);
const initialState = { user: { name: 'jevon' }, roomInfo: {}, playerStocks: {} };
// aconst store = mockStore(initialState);

it('BuyModalMain renders correctly', () => {
  const wrapper = shallow(<BuyModalMain stock='LFIN' roomStocks={{ LFIN: [10] }} user={{ cash: 1000 }}/>);

  expect(wrapper).toMatchSnapshot();
});


it('SellSharesModal renders correctly', () => {
  const wrapper = shallow(<SellSharesModal />);

  expect(wrapper).toMatchSnapshot();
});
