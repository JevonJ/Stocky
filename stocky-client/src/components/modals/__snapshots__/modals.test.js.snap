// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuyModalMain renders correctly 1`] = `
<Modal
  autoFocus={true}
  backdrop={true}
  backdropTransition={
    Object {
      "mountOnEnter": true,
      "timeout": 150,
    }
  }
  centered={false}
  fade={true}
  id="BuyModal"
  isOpen={false}
  keyboard={true}
  modalTransition={
    Object {
      "timeout": 300,
    }
  }
  onClosed={[Function]}
  onOpened={[Function]}
  role="dialog"
  toggle={[Function]}
  zIndex={1050}
>
  <ModalHeader
    closeAriaLabel="Close"
    style={
      Object {
        "color": "black",
      }
    }
    tag="h5"
    wrapTag="div"
  >
    You are going to buy
  </ModalHeader>
  <ModalBody
    className="mbody"
    style={
      Object {
        "color": "black",
      }
    }
    tag="div"
  >
    <ListGroup
      tag="ul"
    >
      <ListGroupItem
        tag="li"
      >
        Company Symbol: 
        LFIN
      </ListGroupItem>
      <ListGroupItem
        tag="li"
      >
        Unit Price: 
        10
      </ListGroupItem>
      <ListGroupItem
        tag="li"
      >
        <FormGroup
          tag="div"
        >
          Number of Stocks
          <Input
            max={100}
            min={1}
            name="qauntity"
            onChange={[Function]}
            placeholder="Number of Stocks"
            type="number"
            value="1"
          />
        </FormGroup>
      </ListGroupItem>
      <ListGroupItem
        tag="li"
      >
        Price: 
        10
      </ListGroupItem>
    </ListGroup>
  </ModalBody>
  <ModalFooter
    tag="div"
  >
    <Button
      color="success"
      onClick={[Function]}
      tag="button"
    >
      Buy
    </Button>
    <Button
      color="danger "
      onClick={[Function]}
      tag="button"
    >
      Not interested
    </Button>
  </ModalFooter>
</Modal>
`;

exports[`SellSharesModal renders correctly 1`] = `
<Modal
  autoFocus={true}
  backdrop={true}
  backdropTransition={
    Object {
      "mountOnEnter": true,
      "timeout": 150,
    }
  }
  centered={false}
  fade={true}
  id="SellModal"
  isOpen={false}
  keyboard={true}
  modalTransition={
    Object {
      "timeout": 300,
    }
  }
  onClosed={[Function]}
  onOpened={[Function]}
  role="dialog"
  toggle={[Function]}
  zIndex={1050}
>
  <ModalHeader
    closeAriaLabel="Close"
    style={
      Object {
        "color": "black",
      }
    }
    tag="h5"
    wrapTag="div"
  >
    You are going to sell
  </ModalHeader>
  <ModalBody
    className="mbody"
    style={
      Object {
        "color": "black",
      }
    }
    tag="div"
  >
    <ListGroup
      tag="ul"
    >
      <ListGroupItem
        tag="li"
      >
        Company Symbol: 
      </ListGroupItem>
      <ListGroupItem
        tag="li"
      >
        <FormGroup
          tag="div"
        >
          Number of Stocks
          <Input
            min={1}
            name="qauntity"
            onChange={[Function]}
            placeholder="Number of Stocks"
            type="number"
            value="1"
          />
        </FormGroup>
      </ListGroupItem>
      <ListGroupItem
        tag="li"
      >
        Price: 
      </ListGroupItem>
      <ListGroupItem
        tag="li"
      >
        Profit/Loss:
         
      </ListGroupItem>
    </ListGroup>
  </ModalBody>
  <ModalFooter
    tag="div"
  >
    <Button
      color="success"
      onClick={[Function]}
      tag="button"
    >
      Sell
    </Button>
    <Button
      color="danger "
      onClick={[Function]}
      tag="button"
    >
      Not interested
    </Button>
  </ModalFooter>
</Modal>
`;
